racket-cas/racket-cas.rkt

```diff
@@ -1318,28 +1323,45 @@
 
 ; A term of the form
 ;        @i        or
-;   (*   @i u ...) or
-;   (* r @i u ...)
+;   (*   u ... @i) or
+;   (* r u ... @i)
 ; where r is a real numbers, is called an *imaginary term*.
```

Comment:
The output looks "better" when @i is the last element in a product - but ... the price is too high efficiency wise

The old `imaginary-term?` used the match pattern `(list* '*   '@i _)` and the new uses the pattern `(list '*   _ ... '@i)`.
For the old pattern the matcher can look at the two first elements - and for the new it needs to look at the last element.
This turns `imaginary-term?` from an O(1) function into an O(n) function - and that's problematic.

The price is paid not only when complex expressions are in use but also for purely real expressions - so the system as whole
will become slower, also when complex numbers aren't used..

That is: the representation (with @i as the first symbol in a product) was chosen to make `imaginary-term?` fast.

Reply:
Thanks for the comments. I did not realize this before.
Then is it okay to put @i as the first item of Sum / Product? (uniform usage of <<=?)

If we want to put @i/Imaginary-term as the last item of Sum, but as first item of Product,
we need to handle @I specially and cannot rely on symbol<<? directly.

Reply:
Then is it okay to put @i as the first item of Sum / Product? 
Not with the current definition of normalized expressions. 
But it could be okay, if we the normalization rules change.
I am reluctant to change them though - there are a lot of subtle 
dependencies on the normalization order.

What is the problem you want to solve by changing the definition of a normalized expression?

Reply:
yes, this is exactly what I was trying to say by “ put @i as the first item of Sum / Product”.

Reply:
Fixed.

Reply:
Hi. I just committed the changes we talked about. Now @i is first in both sums and products.
I have added/modified match patterns (Complex a b) which matches and binds a to the real part and b to the imaginary part. Also the match pattern (ImaginaryTerm u) matches only imaginary terms.

Some of your pull requests now say "conflicts must be resolved". Sorry about that.